src/pymor/reductors/sobt.py

```diff
@@ -0,0 +1,390 @@
+
+
+class GenericSOBTpvReductor(GenericPGReductor):
+    """Generic Second-Order Balanced Truncation position/velocity reductor.
+
+    See [RS08]_.
+
+    .. [RS08] T. Reis and T. Stykel,
+              Balanced truncation model reduction of second-order
+              systems,
+              Math. Comput. Model. Dyn. Syst., 2008, 14(5), 391-406
+
+    Parameters
+    ----------
+    d
+        The system which is to be reduced.
+    """
+    def __init__(self, d):
+        assert isinstance(d, SecondOrderSystem)
+        self.d = d
+        self.V = None
+        self.W = None
+
+    def gramians(self):
+        """Returns gramians."""
+        raise NotImplementedError
+
+    def projection_matrices_and_singular_values(self, r, gramians):
+        """Returns projection matrices and singular values."""
+        raise NotImplementedError
+
+    def reduce(self, r, projection='bfsr'):
+        """Reduce using SOBTp.
+
+        Parameters
+        ----------
+        r
+            Order of the reduced model.
+        projection
+            Projection method used:
+
+                - `'sr'`: square root method
+                - `'bfsr'`: balancing-free square root method (default,
+                    since it avoids scaling by singular values and
+                    orthogonalizes the projection matrices, which might
+                    make it more accurate than the square root method)
+                - `'biorth'`: like the balancing-free square root
+                    method, except it biorthogonalizes the projection
+                    matrices
+
+        Returns
+        -------
+        rd
+            Reduced system.
+        """
+        assert 0 < r < self.d.n
+        assert projection in ('sr', 'bfsr', 'biorth')
+
+        # compute all necessary Gramian factors
+        gramians = self.gramians()
+
+        if r > min(len(g) for g in gramians):
+            raise ValueError('r needs to be smaller than the sizes of Gramian factors.')
+
+        # compute projection matrices and find the reduced model
+        self.V, self.W, singular_values = self.projection_matrices_and_singular_values(r, gramians)
+        if projection == 'sr':
+            alpha = 1 / np.sqrt(singular_values[:r])
+            self.V.scal(alpha)
+            self.W.scal(alpha)
+            self.biorthogonal_product = None
+        elif projection == 'bfsr':
+            self.V = gram_schmidt(self.V, atol=0, rtol=0)
+            self.W = gram_schmidt(self.W, atol=0, rtol=0)
+            self.biorthogonal_product = None
+        elif projection == 'biorth':
+            self.V, self.W = gram_schmidt_biorth(self.V, self.W, product=self.d.M)
+            self.biorthogonal_product = 'M'
+
+        rd = super().reduce()
+
+        return rd
+
+    extend_source_basis = None
+    extend_range_basis = None
+
+
+class SOBTpReductor(GenericSOBTpvReductor):
+    """Second-Order Balanced Truncation position reductor.
+
+    See [RS08]_.
```

Comment:
Would be better to repeat the reference, so this makes sense when just looking at the docstring of SOBTpReductor.